### ISSUE TITLE
No output with option --depends in version 2.7.x

### ISSUE THREAD

**Issue by gkraser** [less/less.js#2925]
If run lessc 2.7.x with option '--depends' (lessc in.less out.css --depends), depends writen in console, output file - not writed.
In version 2.6.1 - all ok.

**Comment by seven-phases-max** [comment]
This was implemented in #2830. So what is the use-case of using both ("depends list" and "css") simultaneously?

**Comment by jdufresne** [comment]
The first run of the Makefile, CSS files _must_ be generated as they do not yet exist. As a byproduct of producing the CSS files, the dependency files are created simultaneously. This avoid shelling out to two processes when one will do. Thus, theoretically, parsing the less files needs to happen only once. As the files are rebuilt, the dependencies will also be updated.

As an example, this is how [GCC works](https://gcc.gnu.org/onlinedocs/gcc-2.95.3/gcc_2.html#SEC11). 

See the article [Autodependencies with GNU make](http://scottmcpeak.com/autodepend/autodepend.html).

**Comment by seven-phases-max** [comment]
I think this is basically leads us to https://github.com/less/less.js/issues/2476#issuecomment-76669260.
I dunno if `lessc` is ever going to replicate the antihuman `gcc` configurabilty (yet another zillion of zillions lines of code).
So I guess we need to try to reduce the decision just to: "what is more important - the timing of the first `lessc` run when we do need both (i.e. your case) OR the timing of further runs when no css is often needed (@gtalusan  case)".

**Comment by de-vri-es** [comment]
When considering `make`, you get the best possible processing time exactly when  you produce the dependencies and the output in one run. There simply is no need to generate *just* the dependencies. Right now I'm forced to process the file twice if I want dependency tracking. And both runs take about the same time.

Maybe there are use cases outside of `make` where having dependencies without actual output makes sense, but for `make` #2830 is a huge setback.

**Comment by seven-phases-max** [comment]
Just run `lescc` w/o `--depends` first (note the only reason for #2830 was performance, i.e. you still got the same `old behavior` = `lessc-generate-css-only + less-generate-dependencies-only`).

So the only drawback is that you need additional:
```
if (!css-exists)
    generate it;
```
in your scripts. Well, I guess this is survivable.

---
<sub>And I still don't get the use-case of having the css output before the eps. I.e. if your already got your CSS *why* do you need that dependencies *its* all?</sub>

**Comment by seven-phases-max** [comment]
added "Consider Closing" due to lack of clear use-case description.

**Comment by de-vri-es** [comment]
The description in #2830 describes that `--depends` **improved** performance, with the only downside being that **he** now ran `lessc` twice. That was their mistake, they should just generate both the dependency files and the output in a single pass. There is no downside to that, and it avoids running `lessc` twice.

With the new behaviour, the only way to gain any benefit from `--depends`  is to run `lessc` twice, which takes twice as long as running it once. That really is not an improvement.

I'm working on a few more improvements to dependency tracking. I'll submit a PR soon.

**Comment by de-vri-es** [comment]
In case it is not clear: there is no point in having dependency information before the generated output. The dependency tracking makes sure that `make` only invokes `lessc` when one of the source files is newer than the output file.

If there is no output file yet, `make` must always run `lessc` so there is no value in having the dependency information before the real output.

**Comment by seven-phases-max** [comment]
@de-vri-es

So the only use of these generated dependencies is to trigger recompilation when they changed?
In that case generating css+deps simultaneosly indeed makes more sense. Then I'm more in favour of reverting #2830 (as there the use-case assumes your dependencies have to be regenerated after the first pass thus the first css output is supposed to be invalid anyway.. Or somthing like that).

**Comment by de-vri-es** [comment]
@seven-phases-max: yes, that is what `make` can use the generated output for.

Note that I submitted #3087, which makes the default behaviour what we had before #2830, but also adds a `--no-write` flag to suppress regular CSS output. In addition that PR adds `-MP`, `-MT` and `-MF` flags with the same meaning as GCC (and they have friendlier longer names as well).

**Comment by de-vri-es** [comment]
Also, I guess that there might be use cases outside of `make`, but the option description does talk about makefiles specifically. In any case, with `--no-write` it would still be possible to suppress CSS output.

**Comment by kevinpeno** [comment]
Is this still an issue that needs addressing? If so, what is the expected result for an incoming PR to solve?

**Comment by seven-phases-max** [comment]
If I understand it right #3087 had the behaviour we were agreed with - it was not merged just because it forced four or five more bloatware options.

**Comment by kevinpeno** [comment]
@seven-phases-max so basically, make `--depends` options and add the `--no-css` option as described. Ignore the rest of the additional options. Am I correct in this understanding?

**Comment by seven-phases-max** [comment]
>add the --no-css option as described

Can we invent something w/o adding new options?

**Comment by kevinpeno** [comment]
@seven-phases-max I can implement it however you like, I was just trying to digest the PR you mentioned into an actionable new PR to help close this issue. If you'd like me to just complete item 1, that's fine.

**Comment by seven-phases-max** [comment]
@kevinpeno Since I do not use this options myself, my only wish is to not burden the further maintenance by adding surplus options (then tests, then documentation, then issue reports etc.). So *for me* something like https://github.com/less/less.js/issues/2925#issuecomment-316849640 would do the trick. 
But since it's *you* going to put efforts please feel free to suggest anything if you feel just reverting #2830 is not enough (well, *reverting* #2830 as such is impossible as there're too many commits since then so I assume "restoring" pre-#2830 behaviour of course).

**Comment by de-vri-es** [comment]
I had a need for more flexible dependency tracking again today, so I started this: https://github.com/de-vri-es/lessc-make

Still a work in progress, but it should already work for simple use cases with source maps and dependency tracking. No plugins or other compiler options yet.

**Comment by ritschwumm** [comment]
i too would be very much in favor of gathering dependencies (ideally written to a file instead of stdout) and writing css output at the same time.

my use case is a simple build script where (to save time) i want to skip running lessc when its inputs have not changed. without  getting both pieces of data (css and input files) at the same, i don't think i can: 

if i want to know about the inputs, i have to run `lessc --depends` one way or another -- and since `--depends` is not (much) faster than full css generation, i have not won anything. 

on a side note, it's feels a bit weird that while --depends disables writing css files, it does _not_ disable writing source maps...

**edit**

thinking about this some more, i came to the conclusion that fiddling with how and when a makefile is written is trying to solve the wrong problem, so how about this:

```lessc --state target/less-state.json```

with this option, lessc would writes a list of its inputs and their mtime (or checksum?) into `state.json` after each compilation.

under certain circumstances lessc would skip compilation and just exit with RC 0:
- all output files already exist
- state.json exists
- all input files mentioned in state.json still have the same mtime (or checksum) they had when state.json was created

**Comment by matthew-dean** [comment]
@ritschwumm You're commenting on a very old issue. I'll admit I don't know anything about this option and it's not well-documented in Less documentation. https://lesscss.org/usage/

Would you be willing to create a new issue that states:

- what the current behavior is

- what you would expect or what you would like to have happen

- why? (the use case)

**Comment by matthew-dean** [comment]
@less/maintainers we probably could use some regression tests for each CLI option, because I'm not sure all of the Less options even have test coverage

**Comment by ritschwumm** [comment]
@matthew-dean i wasn't quite sure yet whether "skip compilation for unchanged sources" and "create a makefile" actually different (albeit related) problems, or just two sides of the same coin, and hoped for someone else chiming in.

i now think it's the former and "skip compilation" would better be implemented as a separate feature, so yes, i can open an issue for that. i might try to cook up a PR myself, but i have to admit i have a bit of a hard time reading plain javascript without any types.

**Comment by matthew-dean** [comment]
@ritschwumm 

> i have a bit of a hard time reading plain javascript without any types.

The JavaScript code for Less is incredibly hard to reason about. It's one of the reasons I've been working on re-writing it from scratch. I asked the Cursor AI chat how long it might take to even just add types to the existing code so that all the files are type-checkable and it estimated 3-6 months :/

**Comment by ritschwumm** [comment]
@matthew-dean when i first looked into the code, i had the strong urge do start migrating everything to ts. wouldn't be the first time i do this... but imagine the PR, not exactly the kind i would want to review :)

**Comment by ritschwumm** [comment]
see #4361 for more about avoiding compilation
